### ISSUE TITLE
Stop building EME by default.

### ISSUE THREAD

**Issue by wolfbeast** [MoonchildProductions/moebius#177]
With our initial platform release out the door, EME is a default-on option that needs to go.
Building EME needs to be made opt-in, and be excluded from Basilisk releases.

- Flip the configure default, but enable for Basilisk in mozconfig
- Add a release note with M2 for people building from source
- ~~Make sure EME is built ON with internal test builds occasionally to prevent bitrot~~

**Comment by wolfbeast** [comment]
Keeping this enabled (through configure) in Basilisk releases will keep it from needing special attention for the EME components and Firefox users wanting a 1:1 will expect it to be present. Basilisk should be a drop-in replacement as much as possible for those users.

**Comment by wolfbeast** [comment]
There's actually nothing to flip, here. The option is --enable-eme=value, it's just set to default to widevine (the only supported CDM at this point) on `app=browser`; removing that should be sufficient to require the mozconfig setting or it won't build.

**Comment by wolfbeast** [comment]
OK, so, diving into this a bit more: apparently there is no way to actually exclude it at build time. That was removed as part of landing Firefox 52, where the build-time configure option `MOZ_EME` was removed in favor of controlling everything with preferences. That was done in [Bug1300654](https://bugzilla.mozilla.org/show_bug.cgi?id=1300654). If we want to reinstate build-time inclusion/exclusion of EME code in our binaries, then this needs to be reverted.

The question becomes then: Is this a priority? Do we want to keep it simpler and have a single build path that we don't need to test alternatives of separately, or do we want to remove the otherwise completely-disabled EME code because it's based on a pref? Firefox's "EME-free" versions are in fact doing the former and shipping a distribution.ini that flips the `media.eme.enabled` pref to false (our default) and doesn't actually change the build/inclusion. The code doesn't harm anyone, is fully open-sourced, isn't hooked into anything in a dangerous way as far as I can tell, and any other applications building from the UXP tree can control this system completely by switching prefs:
```JavaScript
// Encrypted media extensions.
// EME is visible but disabled by default. This is so that the
// "Play DRM content" checkbox in the Firefox UI is unchecked by default.
// DRM requires downloading and installing proprietary binaries, which
// users didn't necessarily opt-in to by installing the browser.
// The first time a site using EME is encountered, the user will be
// prompted to enable DRM, whereupon the EME plugin binaries will be
// downloaded if permission is granted.
pref("media.eme.enabled", false);
pref("media.eme.apiVisible", true);

// Whether we should run a test-pattern through EME GMPs before assuming they'll
// decode H.264.
pref("media.gmp.trial-create.enabled", true);

// Note: when media.gmp-*.visible is true, provided we're running on a
// supported platform/OS version, the corresponding CDM appears in the
// plugins list, Firefox will download the GMP/CDM if enabled, and our
// UI to re-enable EME prompts the user to re-enable EME if it's disabled
// and script requests EME. If *.visible is false, we won't show the UI
// to enable the CDM if its disabled; it's as if the keysystem is completely
// unsupported.
pref("media.gmp-widevinecdm.visible", true);
pref("media.gmp-widevinecdm.enabled", false);
```

Setting the `apiVisible` and/or `media.gmp-*.visible` prefs to false will effectively completely disable EME, and no CDM will ever be downloaded. Removing visbility of the API will also automatically remove the UI component in Options in Basilisk, at least.

@mattatobin I know it's not an idyllic solution, but it's effective nonetheless. EME code can always be made conditional at build time and removed later when/if it becomes prudent to do so; I just don't really see the point in making it all build-time conditional right now since there is full control. Let me know if you have any objections to a `wontfix` here.

**Comment by mattatobin** [comment]
Guess I will work it out next year.

**Comment by wolfbeast** [comment]
I can at least make the EME settings more harsh and disable API visibility based on building with --enable-eme or not, for now. Backing out the FF52 bug that removed build conditionals doesn't seem to be too difficult, so anyone could do it, really, as long as they are careful about it. it seems to have been a regression for Mozilla's convenience more than anything else.

**Comment by wolfbeast** [comment]
Looks like the actual rationale behind this on "EME-free" when cutting through the BS is the following: 

> This is the behaviour that we want for the EME-Free repacks; we don't want users to be tricked into downloading these builds and then finding out that they're stuck not being able to make DRM protected video work; there should be a path to make EME work, and that's what these builds have.

So yes, it's very much by design at Mozilla to have it always-on or at least always-available, even on "EME-free" which becomes not-really-EME-free as a result. I guess the people maintaining that minimal fork didn't want to fight it.

**Comment by mattatobin** [comment]
The other issue is if the pref was enabled in Pale Moon on UXP then it would be in a very odd situation because there are no UI bits in our browser code to deal with it.. It would completely undefined.

**Comment by wolfbeast** [comment]
I'll work on this to have it be build-time; this is a blocker for the intended build system changes as discussed with @mattatobin. Should be done before the holidays.

**Comment by wolfbeast** [comment]
It's more complicated than it looked at first glance.

Basically, Mozilla broke `--disable-eme` by detaching `GMPDecryptorProxyCallback` from GMP and making it a re-usable API through refactoring. The bug that was raised resulted in working `MOZ_EME` completely out of the build system, which was (see above) a desired move anyway.

I _could_ unrefactor things and re-attach the CDM proxy callbacks, but to do so I'd introduce a bunch of bugs that were fixed and would have to be re-fixed (but then on a different base) for EME to work according to spec. I really don't see a point in purposefully having to break and then rewrite a working setup, so I think I'll have to call it here, and keep the pref-controlled setup, but then with hard controls on the API based on `MOZ_EME`. 
Effectively there won't be an API if I fix it up the way I intend, so even if compiled, the EME code will effectively be NOP. It will still be controlled at build-time then, but there will be some EME binary parts in the `disable` scenario that are built because they have become dependencies for other code in the platform, most notably the media-plugins (GMP). It'll have to be a compromise.

Compromises will have to be made in terms of what actually gets compiled and included, although MOZ_EME can easily be used to make the API completely NOP when configured as --disable-eme (the default) without impacting the EME functionality.

**Comment by wolfbeast** [comment]
The current state is about as good as it'll get for now.
Closing.
